README.md

```diff
@@ -1,9 +1,509 @@
+Persistent [Adaptive Radix Trees](https://db.in.tum.de/~leis/papers/ART.pdf).
 
-The database is built on top of Persistent Adative Radix Trees (implemented in part/). It supports
-multi-version concurrency control and transactional cross-table write transaction. Changes to the database
-can be watched at fine-granularity which allows reacting to changes in the database.
+StateDB is/supports:
+
+* In-memory. Objects and indexes are stored in main memory and not on disk.
+  This makes it easy to store and index any Go data type.
+
+* Multi-Version Concurrency Control (MVCC). Both objects and indexes are immutable
+  and objects are versioned. A read transaction has access to an immutable snapshot
+  of the data.
+
+* Cross-table write transactions. Write transactions lock the requested tables and
+  allow modifying objects in multiple tables as a single atomic action. Transactions
+  can be aborted to throw away the changes.
+
+* Multiple indexes. A table may have one or more indexers for objects, with each
+  indexer returning zero or more keys. Indexes can be unique or non-unique.
+  A non-unique index is a concatenation of the primary and secondary keys.
+
+* Watch channels. Changes to the database can be watched at fine-granularity via
+  Go channels that close when a relevant part of the database changes. This is
+  implemented by having a Go channel at each of the radix tree nodes. This enables
+  watching an individual object for changes, a key prefix, or the whole table.
+
+## Warning! Immutable data! Read this!
+
+To support lockless readers and transactionality StateDB relies on both the indexes
+and the objects themselves being immutable. Since in Go you cannot declare fields
+`const` we cannot stop mutation of public fields in objects. This means that care
+must be taken with objects stored in StateDB and not mutate objects that have been
+inserted into it. This means both the fields directly in the object and everything
+referenced from it, e.g. a map field must not be modified, but must be cloned first!
+
+StateDB has a check in `Insert()` to validate that if an object is a pointer then it
+cannot be replaced with the same pointer, but that at least a shallow clone has been
+made. This of course doesn't extend to references within the object.
+
+For "very important objects", please consider storing an interface type instead that
+contains getter methods and a safe way of mutating the object, e.g. via the builder
+pattern or a constructor function. 
+
+Also prefer persistent/immutable data structures within the object to avoid expensive
+copying on mutation. The `part` package comes with persistent `Map[K]V` and `Set[T]`.
+
+With this disclaimer out of the way, let's dive into how to use StateDB.
+
+## StateDB usage
+
+StateDB can be used directly as a normal library, or as a Hive Cell. For example
+usage as part of Hive, see `reconciler/example`. Here we show a standalone example.
```

Comment:
> standalone example

Additionally maybe a section on how to use it _with_ hive?

Reply:
I'll leave that to another day.